prettyPrinters/providers.py

```diff
@@ -266,3 +266,83 @@ def update(self):
     def has_children(self):
         # type: () -> bool
         return True
+
+
+def StdRcSummaryProvider(valobj, dict):
+    # type: (SBValue, dict) -> str
+    assert valobj.GetNumChildren() == 3
+
+    strong = valobj.GetChildMemberWithName("strong").GetValueAsUnsigned()
+    weak = valobj.GetChildMemberWithName("weak").GetValueAsUnsigned()
+    return "strong={}, weak={}".format(strong, weak)
+
+
+class StdRcSyntheticProvider:
+    """Pretty-printer for alloc::rc::Rc<T> and alloc::sync::Arc<T>
+
+    struct Rc<T> { ptr: NonNull<RcBox<T>>, ... }
+    struct NonNull<T> { pointer: NonZero<*const T> }
+    struct NonZero<T>(T)
+    struct RcBox<T> { strong: Cell<usize>, weak: Cell<usize>, value: T }
+    struct Cell<T> { value: UnsafeCell<T> }
+    struct UnsafeCell<T> { value: T }
+
+    struct Arc<T> { ptr: NonNull<ArcInner<T>>, ... }
+    struct NonNull<T> { pointer: NonZero<*const T> }
+    struct NonZero<T>(T)
+    struct ArcInner<T> { strong: atomic::AtomicUsize, weak: atomic::AtomicUsize, data: T }
+    struct AtomicUsize { v: UnsafeCell<usize> }
+    struct UnsafeCell<T> { value: T }
+    """
+
+    def __init__(self, valobj, dict, is_atomic=False):
+        # type: (SBValue, dict, bool) -> StdRcSyntheticProvider
+        self.valobj = valobj
+
+        self.ptr = self.valobj.GetChildMemberWithName("ptr").GetChildMemberWithName("pointer").GetChildAtIndex(0)
+        self.value = self.ptr.GetChildMemberWithName("data" if is_atomic else "value")
+
+        self.strong = self.ptr.GetChildMemberWithName("strong").GetChildAtIndex(0).GetChildMemberWithName("value")
+        self.weak = self.ptr.GetChildMemberWithName("weak").GetChildAtIndex(0).GetChildMemberWithName("value")
+
+        process = valobj.GetProcess()
+        self.endianness = process.GetByteOrder()
+        self.pointer_size = process.GetAddressByteSize()
+        self.count_type = valobj.GetType().GetBasicType(eBasicTypeUnsignedLong)
+
+    def num_children(self):
+        # type: () -> int
+        return 3
+
+    def get_child_index(self, name):
+        # type: (str) -> int
+        if name == "strong":
+            return 0
+        if name == "weak":
```

Comment:
Do we really need `strong` and `weak` children at all? For example, formatter for C++ `shared_ptr` doesn't provide such children, only add them into summary

Reply:
Yes, C++ pretty-printer uses the interesting trick:
```
def num_children(self):
    return 1

def get_child_at_index(self, index):
    if index == 0: ...
    if index == 1: ...
    if index == 2: ...
```
So the synthtetic provider doesn't show the 1st and 2nd children, but summary provider can access (and show) these.

I will use it too.